flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/nodes/exec/common/CommonExecSink.java

```diff
@@ -260,76 +191,135 @@ public DynamicTableSinkSpec getTableSinkSpec() {
      * the parallelism is provided, otherwise it uses parallelism of input transformation.
      */
     private int deriveSinkParallelism(
-            ParallelismProvider parallelismProvider, int inputParallelism) {
-        final Optional<Integer> parallelismOptional = parallelismProvider.getParallelism();
-        if (parallelismOptional.isPresent()) {
-            int sinkParallelism = parallelismOptional.get();
-            if (sinkParallelism <= 0) {
-                throw new TableException(
-                        String.format(
-                                "Table: %s configured sink parallelism: "
-                                        + "%s should not be less than zero or equal to zero",
-                                tableSinkSpec.getObjectIdentifier().asSummaryString(),
-                                sinkParallelism));
-            }
-            return sinkParallelism;
-        } else {
-            // use input parallelism if not specified
+            Transformation<RowData> inputTransform, SinkRuntimeProvider runtimeProvider) {
+        final int inputParallelism = inputTransform.getParallelism();
+        if (!(runtimeProvider instanceof ParallelismProvider)) {
             return inputParallelism;
         }
+        final ParallelismProvider parallelismProvider = (ParallelismProvider) runtimeProvider;
+        return parallelismProvider
+                .getParallelism()
+                .map(
+                        sinkParallelism -> {
+                            if (sinkParallelism <= 0) {
+                                throw new TableException(
+                                        String.format(
+                                                "Invalid configured parallelism %s for table '%s'.",
+                                                sinkParallelism,
+                                                tableSinkSpec
+                                                        .getObjectIdentifier()
+                                                        .asSummaryString()));
```

Comment:
This is a complete nit, but I don't really understand why we're going through the trouble of calling a method (rather than relying on toString) to print a potentially incorrect representation rather than a properly escaped one.

Reply:
Let's not start this discussion now. It seems the backticks are hard to read in strings. We could have used `asSummaryString` in `toString` but now it is kind of too late I guess.